File: Managed Source/Arsenal.ImageMounter/IO/ConsoleProgressBar.vb
Imports System.Security
Imports System.Security.Permissions
Imports System.Text
Imports System.Threading

#Disable Warning IDE0079 ' Remove unnecessary suppression
#Disable Warning SYSLIB0003 ' Type or member is obsolete

Namespace IO

    <SecuritySafeCritical>
    Public Class ConsoleProgressBar
        Implements IDisposable

        Public ReadOnly Property Timer As New Timer(AddressOf Tick)

        Public ReadOnly Property CurrentValue As Integer

        Private ReadOnly updateFunc As Func(Of Double)

        <SecuritySafeCritical>
        Private Sub New(update As Func(Of Double))
            updateFunc = update
            CreateConsoleProgressBar()
        End Sub

        Public Sub New(dueTime As Integer, period As Integer, update As Func(Of Double))
            Me.New(update)

            Timer.Change(dueTime, period)

        End Sub

        Public Sub New(dueTime As TimeSpan, period As TimeSpan, update As Func(Of Double))
            Me.New(update)

            Timer.Change(dueTime, period)

        End Sub

        <SecuritySafeCritical>
        Private Sub Tick(o As Object)

            Dim newvalue = updateFunc()

            If newvalue <> 1D AndAlso CInt(100 * newvalue) = _CurrentValue Then
                Return
            End If

            UpdateConsoleProgressBar(newvalue)

        End Sub

        <SecuritySafeCritical>
        <SecurityPermission(SecurityAction.Demand, Flags:=SecurityPermissionFlag.AllFlags)>
        Public Shared Sub CreateConsoleProgressBar()

            If NativeFileIO.UnsafeNativeMethods.GetFileType(NativeFileIO.UnsafeNativeMethods.GetStdHandle(NativeFileIO.StdHandle.Output)) <> NativeFileIO.Win32FileType.Character Then
                Return
            End If

            Dim row As New StringBuilder(Console.WindowWidth)

            row.Append("["c)

            row.Append("."c, Math.Max(Console.WindowWidth - 3, 0))

            row.Append("]"c)

            row.Append(vbCr(0))

            SyncLock ConsoleSync

                Console.ForegroundColor = ConsoleProgressBarColor

                Console.Write(row.ToString())

                Console.ResetColor()

            End SyncLock

        End Sub

        <SecuritySafeCritical>
        <SecurityPermission(SecurityAction.Demand, Flags:=SecurityPermissionFlag.AllFlags)>
        Public Shared Sub UpdateConsoleProgressBar(value As Double)

            If NativeFileIO.UnsafeNativeMethods.GetFileType(NativeFileIO.UnsafeNativeMethods.GetStdHandle(NativeFileIO.StdHandle.Output)) <> NativeFileIO.Win32FileType.Character Then
                Return
            End If

            If value > 1D Then
                value = 1D
            ElseIf value < 0 Then
                value = 0D
            End If

            Dim currentPos = CInt((Console.WindowWidth - 3) * value)

            Dim row As New StringBuilder(Console.WindowWidth)

            row.Append("["c)

            row.Append("="c, Math.Max(currentPos, 0))

            row.Append("."c, Math.Max(Console.WindowWidth - 3 - currentPos, 0))

            Dim percent = $" {(100 * value):0} % "

            Dim midpos = (Console.WindowWidth - 3 - percent.Length) >> 1

            If midpos > 0 AndAlso row.Length >= percent.Length Then

                row.Remove(midpos, percent.Length)

                row.Insert(midpos, percent)

            End If

            row.Append("]"c)

            row.Append(vbCr(0))

            SyncLock ConsoleSync

                Console.ForegroundColor = ConsoleProgressBarColor

                Console.Write(row.ToString())

                Console.ResetColor()

            End SyncLock

        End Sub

        Public Shared Sub FinishConsoleProgressBar()

            UpdateConsoleProgressBar(1D)

            Console.WriteLine()

        End Sub

        Public Shared Property ConsoleProgressBarColor As ConsoleColor = ConsoleColor.Cyan

        Private disposedValue As Boolean ' To detect redundant calls

        ' IDisposable
        Protected Overridable Sub Dispose(disposing As Boolean)
            If Not disposedValue Then
                If disposing Then
                    ' TODO: dispose managed state (managed objects).
                    _Timer.Dispose()
                    FinishConsoleProgressBar()

                End If

                ' TODO: free unmanaged resources (unmanaged objects) and override Finalize() below.

                ' TODO: set large fields to null.
                _Timer = Nothing
            End If
            disposedValue = True
        End Sub

        ' TODO: override Finalize() only if Dispose(disposing As Boolean) above has code to free unmanaged resources.
        Protected Overrides Sub Finalize()
            ' Do not change this code.  Put cleanup code in Dispose(disposing As Boolean) above.
            Dispose(False)
            MyBase.Finalize()
        End Sub

        ' This code added by Visual Basic to correctly implement the disposable pattern.
        Public Sub Dispose() Implements IDisposable.Dispose
            ' Do not change this code.  Put cleanup code in Dispose(disposing As Boolean) above.
            Dispose(True)
            ' TODO: uncomment the following line if Finalize() is overridden above.
            GC.SuppressFinalize(Me)
        End Sub

    End Class

End Namespace

File: Managed Source/Arsenal.ImageMounter/IO/ConsoleSupport/ConsoleProgressBar.cs
using System;
using System.Text;
using System.Threading;

#pragma warning disable CS1591 // Missing XML comment for publicly visible type or member

namespace Arsenal.ImageMounter.IO.ConsoleSupport;

public class ConsoleProgressBar : IDisposable
{
    public Timer Timer { get; }

    public int CurrentValue { get; private set; }

    private readonly Func<double> updateFunc;

    private ConsoleProgressBar(Func<double> update)
    {
        Timer = new(Tick);
        updateFunc = update;
        CreateConsoleProgressBar();
    }

    public ConsoleProgressBar(int dueTime, int period, Func<double> update)
        : this(update)
    {
        Timer = new(Tick);
        Timer.Change(dueTime, period);
    }

    public ConsoleProgressBar(TimeSpan dueTime, TimeSpan period, Func<double> update)
        : this(update)
    {
        Timer = new(Tick);
        Timer.Change(dueTime, period);
    }

    private void Tick(object? o)
    {

        var newvalue = updateFunc();

        if (newvalue != 1d && (int)Math.Round(100d * newvalue) == CurrentValue)
        {
            return;
        }

        UpdateConsoleProgressBar(newvalue);

    }

    public static void CreateConsoleProgressBar()
    {

        if (Console.IsOutputRedirected)
        {
            return;
        }

        var row = new StringBuilder(Console.WindowWidth);

        row.Append('[');

        row.Append('.', Math.Max(Console.WindowWidth - 3, 0));

        row.Append("]\r");

        lock (ConsoleSupport.ConsoleSync)
        {
            Console.ForegroundColor = ConsoleProgressBarColor;

            Console.Write(row.ToString());

            Console.ResetColor();
        }
    }

    public static void UpdateConsoleProgressBar(double value)
    {

        if (Console.IsOutputRedirected)
        {
            return;
        }

        if (value > 1d)
        {
            value = 1d;
        }
        else if (value < 0d)
        {
            value = 0d;
        }

        var currentPos = (int)Math.Round((Console.WindowWidth - 3) * value);

        var row = new StringBuilder(Console.WindowWidth);

        row.Append('[');

        row.Append('=', Math.Max(currentPos, 0));

        row.Append('.', Math.Max(Console.WindowWidth - 3 - currentPos, 0));

        var percent = $" {100d * value:0} % ";

        var midpos = Console.WindowWidth - 3 - percent.Length >> 1;

        if (midpos > 0 && row.Length >= percent.Length)
        {

            row.Remove(midpos, percent.Length);

            row.Insert(midpos, percent);

        }

        row.Append("]\r");

        lock (ConsoleSupport.ConsoleSync)
        {
            Console.ForegroundColor = ConsoleProgressBarColor;

            Console.Write(row.ToString());

            Console.ResetColor();
        }
    }

    public static void FinishConsoleProgressBar()
    {
        UpdateConsoleProgressBar(1d);

        Console.WriteLine();
    }

    public static ConsoleColor ConsoleProgressBarColor { get; set; } = ConsoleColor.Cyan;

    private bool disposedValue; // To detect redundant calls

    // IDisposable
    protected virtual void Dispose(bool disposing)
    {
        if (!disposedValue)
        {
            if (disposing)
            {
                // TODO: dispose managed state (managed objects).
                Timer.Dispose();
                FinishConsoleProgressBar();

            }

            // TODO: free unmanaged resources (unmanaged objects) and override Finalize() below.

            // TODO: set large fields to null.
        }

        disposedValue = true;
    }

    // TODO: override Finalize() only if Dispose(disposing As Boolean) above has code to free unmanaged resources.
    ~ConsoleProgressBar()
    {
        // Do not change this code.  Put cleanup code in Dispose(disposing As Boolean) above.
        Dispose(false);
    }

    // This code added by Visual Basic to correctly implement the disposable pattern.
    public void Dispose()
    {
        // Do not change this code.  Put cleanup code in Dispose(disposing As Boolean) above.
        Dispose(true);
        // TODO: uncomment the following line if Finalize() is overridden above.
        GC.SuppressFinalize(this);
    }
}